### PR TITLE
Address breaking changes undici@2=>@3

### PR DESCRIPTION
**Pull request by penx**
Address breaking changes introduced in undici@3

https://github.com/nodejs/undici/commit/9df73a1409769d25dab8d4d9dbeb8e11f0260763

#### Checklist

- [x] run `npm run test` ~and `npm run benchmark`~
*missing script: benchmark
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by penx** [comment]
Moved to draft as I think docs need to be updated

**Comment by penx** [comment]
Reviewed docs and I don't see any reference to `undici. requestTimeout`.

Would this be considered a breaking change though? As any upstream users using `undici. requestTimeout` will need to update their code.

**Comment by mcollina** [comment]
Yes, I will bump the major once this lands.

**Comment by mcollina** [comment]
This is missing the bodyTimeout and its relative error.

**Comment by penx** [comment]
I've added the bodyTimeout option and handled the error and pushed, but I think this needs test coverage.

File: test/undici-timeout.js [hunk 1]
```diff
@@ -31,7 +31,7 @@ async function main () {
   instance.register(From, {
     base: `http://localhost:${target.server.address().port}`,
     undici: {
-      requestTimeout: 100
+      headersTimeout: 100
```
Comment (mcollina) [inline test/undici-timeout.js:34]: could you add a copy of this test with `bodyTimeout`  set instead? Thanks

**Review by mcollina — CHANGES_REQUESTED** [review]

File: test/undici-timeout.js (hunk 1 above)
Comment (penx) [reply test/undici-timeout.js:34]: I created a test server that sends the headers before the body as follows:

```js
const target = http.createServer((req, res) => {
  t.pass('request proxied')
  req.on('data', () => undefined)
  req.on('end', () => {
    res.flushHeaders();
    clock.setTimeout(() => {
      res.end(JSON.stringify({ something: 'else' }));
      t.end()
    }, 1000)
  })
})
```

however `await got.get(`http://localhost:${instance.server.address().port}/`, { retry: 0 })` now throws with 

> RequestError: Parse Error: Content-Length can't be present with Transfer-Encoding

Am I right in thinking that the headers are forwarded to the client as soon as they are received, and then when this runs:

https://github.com/penx/fastify-reply-from/blob/bf4793159d1864f36e366c366098e60eccee99bd/index.js#L179-L181

We get an error because the headers have already been sent?

Comment (mcollina) [reply test/undici-timeout.js:34]: Can you include a full stacktrace?

Comment (penx) [reply test/undici-timeout.js:34]:
```
RequestError: Parse Error: Content-Length can't be present with Transfer-Encoding
    at ClientRequest.<anonymous> (/Users/penx/Development/forks/fastify-reply-from/node_modules/got/dist/source/core/index.js:953:111)
    at Object.onceWrapper (node:events:485:26)
    at ClientRequest.emit (node:events:390:22)
    at ClientRequest.origin.emit (/Users/penx/Development/forks/fastify-reply-from/node_modules/@szmarczak/http-timer/dist/source/index.js:39:20)
    at Socket.socketOnData (node:_http_client:509:9)
    at Socket.emit (node:events:378:20)
    at addChunk (node:internal/streams/readable:313:12)
    at readableAddChunk (node:internal/streams/readable:288:9)
    at Socket.Readable.push (node:internal/streams/readable:227:10)
    at TCP.onStreamRead (node:internal/stream_base_commons:190:23)
    at Socket.socketOnData (node:_http_client:502:22)
    at Socket.emit (node:events:378:20)
    at addChunk (node:internal/streams/readable:313:12)
    at readableAddChunk (node:internal/streams/readable:288:9)
    at Socket.Readable.push (node:internal/streams/readable:227:10)
    at TCP.onStreamRead (node:internal/stream_base_commons:190:23)
    at TCP.callbackTrampoline (node:internal/async_hooks:132:14) {
  code: 'HPE_UNEXPECTED_CONTENT_LENGTH',
  timings: {
    start: 1612976101252,
    socket: 1612976101253,
    lookup: 1612976101254,
    connect: 1612976101255,
    secureConnect: undefined,
    upload: 1612976101255,
    response: undefined,
    end: undefined,
    error: 1612976103275,
    abort: undefined,
    phases: {
      wait: 1,
      dns: 1,
      tcp: 1,
      tls: undefined,
      request: 0,
      firstByte: undefined,
      download: undefined,
      total: 2023
    }
   }
 }
```

Comment (mcollina) [reply test/undici-timeout.js:34]: I think you might need to remove `transfer-encoding` from `headers` in https://github.com/fastify/fastify-reply-from/blob/d4729809effa27d7429e41098938c0bd2dca4051/lib/request.js#L114.

Comment (penx) [reply test/undici-timeout.js:34]: This helps, but then the `UND_ERR_BODY_TIMEOUT` can't be intercepted without an unwanted performance hit, as far as I can tell.

It seems to land here:
https://github.com/penx/fastify-reply-from/blob/bf4793159d1864f36e366c366098e60eccee99bd/index.js#L107-L107

with `err` being null, and then here:

https://github.com/penx/fastify-reply-from/blob/bf4793159d1864f36e366c366098e60eccee99bd/index.js#L136-L136

so starts streaming whatever it receives as the body from undici, in this case:

```json
{
   "statusCode": 500,                
   "code": "UND_ERR_BODY_TIMEOUT",   
   "error": "Internal Server Error", 
   "message": "Body Timeout Error",  
} 
```

The only approach to intercept this I guess would be to wait for the stream to complete first. Pretty sure this wouldn't be wanted so I guess we have the unit test expect this response?

Comment (mcollina) [reply test/undici-timeout.js:34]: I'm pretty sure the body you linked is not what is get streamed out from undici. https://github.com/nodejs/undici/blob/master/lib/core/errors.js#L21 as you can see this is an actual error being emitted.

File: index.js [hunk 2]
```diff
@@ -106,7 +106,7 @@ module.exports = fp(function from (fastify, opts, next) {
         if (!this.sent) {
           if (err.code === 'ERR_HTTP2_STREAM_CANCEL' || err.code === 'ENOTFOUND') {
             onError(this, { error: new createError.ServiceUnavailable() })
-          } else if (err instanceof TimeoutError || err.code === 'UND_ERR_REQUEST_TIMEOUT') {
+          } else if (err instanceof TimeoutError || err.code === 'UND_ERR_HEADERS_TIMEOUT' || err.code === 'UND_ERR_BODY_TIMEOUT') {
```
Comment (mcollina) [inline index.js:109]: Do you think `UND_ERR_BODY_TIMEOUT` could happen here? From what I've read it will happen after we have started streaming the response.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: index.js (hunk 2 above)
Comment (penx) [reply index.js:109]: I think you're right, but I'm going to need to read some more of undici's source code to understand the flow here. I'll pick this up tomorrow morning.

Comment (penx) [reply index.js:109]: i.e.

https://github.com/nodejs/undici/blob/c4e237067dbc784d1cefa2b80a3ca5213a3561c8/lib/core/client.js#L731-L735

https://github.com/nodejs/undici/blob/c4e237067dbc784d1cefa2b80a3ca5213a3561c8/lib/core/client.js#L522-L524      ? setTimeout(onBodyTimeout, client[kBodyTimeout], this)
      : null

Comment (penx) [reply index.js:109]: I think more thought is needed here.

If a user sets `bodyTimeout` but not `headersTimeout`, what would they expect the behaviour to be? And what are our options on handling this?

There are a few edge cases that would need to be considered and perhaps tested for:


## Case 1

The headers take longer to respond than `bodyTimeout`, but the subsequent time taken to send the body being less than `bodyTimeout`.

This should complete successfully, the headers can take as long as undici allows them to, `bodyTimeout` is the time taken from the headers being sent to the body being closed (`res.end()`).

I can add a test for this.

## Case 2

The headers are sent quickly, but the body is not sent at all within the `bodyTimeout`

At the moment, we send response code 500 to the client, along with the error from Undici in the body.

https://github.com/penx/fastify-reply-from/blob/bf4793159d1864f36e366c366098e60eccee99bd/index.js#L130-L130

`stream.destroy(<BodyTimeoutError>)`
https://github.com/nodejs/undici/blob/7f3eef73355ced497107023a7af8b987fd01459c/lib/core/util.js#L105

```json
{
  statusCode: 500,
  code: 'UND_ERR_BODY_TIMEOUT',
  error: 'Internal Server Error',
  message: 'Body Timeout Error'
}
```

## Case 3

The headers being sent quickly, but the body being *partially sent* within the `bodyTimeout` period.

```js
const target = http.createServer((req, res) => {
  t.pass('request proxied')
  req.on('data', () => undefined)
  req.on('end', () => {
    res.flushHeaders()
    res.write('test')
    clock.setTimeout(() => {
      res.end()
      t.end()
    }, 1)
  })
})
```

This partially streams a response ("test") to the client and then hits node internals `stream.destroy(<BodyTimeoutError>)` (as in Case 2): 

https://github.com/nodejs/undici/blob/7f3eef73355ced497107023a7af8b987fd01459c/lib/core/util.js#L105

However on this occasion, node internals seem to behave differently because the body has already been partially streamed.

`got.get` throws a "The server aborted pending request" error, and reports that the response code is '200'. `got` does not allow us to access the partially streamed response, as far as I can see.

Confirmed looking at a local proxy that the response code remains as 200 and "test" is sent to the client.

---

I don't think the behaviour of 2 and 3 are great, but I'm going to need to do some more reading to figure out if we have any options to improve them.

Comment (penx) [reply index.js:109]: > Do you think UND_ERR_BODY_TIMEOUT could happen here? From what I've read it will happen after we have started streaming the response.

Perhaps if the response code changes to 500 before we start streaming the body to the client, we could intercept it 🤷 

But as the code stands at the moment, no I don't think `UND_ERR_BODY_TIMEOUT` would be intercepted.

Comment (mcollina) [reply index.js:109]: As for the HTTP spec, headers are alway sent _before_ the body. If they are _after_ the body are called trailers. I don't think this module supports trailers.

Comment (penx) [reply index.js:109]: does that include any response code change that is sent after `flushHeaders` but before anything is written to the body?

Comment (penx) [reply index.js:109]: ah I've been using `res.flushHeaders()` on the fake servers when I should probably be using `res.writeHead(200)` or similar, which changes things a bit

Comment (penx) [reply index.js:109]: ok so in case 2, the headers have not yet been sent to the client, so when an error is received from the target server, and we call `payload.pipe(res)`, something in node internals seems to change the statusCode from 200 to 500.

In case 3, the headers have already been sent to the client, and so node has no choice but to terminate the response.

We may be able to do something in case 2 but I'd need to read the node internals code to see what it's doing when it changes the status to 500.

Comment (mcollina) [reply index.js:109]: Read up on https://developer.mozilla.org/en-US/docs/Web/HTTP/Messages.

TL;DR the order is start lines (status code, etc), headers, body, trailers.

> In case 3, the headers have already been sent to the client, and so node has no choice but to terminate the response (well I guess it could send a trailer as you say, but it doesn't seem to).

Yes. The response has already been sent. You'll get an error on the other side.

Comment (penx) [reply index.js:109]: ok looks like I can improve case 2 by extracting the error handler and setting it when before calling `this.send(res.stream)`

i.e.

```js
 const errorHandler = (err, request, reply) => {
      reply.request.log.warn(err, 'response errored')
        if (!reply.sent) {
          if (err.code === 'ERR_HTTP2_STREAM_CANCEL' || err.code === 'ENOTFOUND') {
            onError(reply, { error: new createError.ServiceUnavailable() })
          } else if (err instanceof TimeoutError || err.code === 'UND_ERR_HEADERS_TIMEOUT' || err.code === 'UND_ERR_BODY_TIMEOUT') {
            onError(reply, { error: new createError.GatewayTimeout() })
          } else {
            onError(reply, { error: createError(500, err) })
          }
        }
    }

    request({ method: req.method, url, qs, headers: requestHeaders, body }, (err, res) => {
      if (err) {
        errorHandler(err, this.request, this)
        return
      }
      this.request.log.info('response received')
      if (sourceHttp2) {
        copyHeaders(
          rewriteHeaders(stripHttp1ConnectionHeaders(res.headers)),
          this
        )
      } else {
        copyHeaders(rewriteHeaders(res.headers), this)
      }
      this.code(res.statusCode)
      if (onResponse) {
        onResponse(this.request.raw, this, res.stream)
      } else {
        this.context.errorHandler = errorHandler
        this.send(res.stream)
      }
    })
    return this
  })
```

Comment (penx) [reply index.js:109]: (note `errorHandler` doesn't fire on case 3 and I'm not sure there's much we could do other than terminating the response, I can add a test case for this)

File: index.js [hunk 3]
```diff
@@ -127,6 +131,7 @@ module.exports = fp(function from (fastify, opts, next) {
       if (onResponse) {
         onResponse(this.request.raw, this, res.stream)
       } else {
+        this.context.errorHandler = errorHandler
```
Comment (mcollina) [inline index.js:134]: you can't do this :(, it will leak memory!

**Review by mcollina — CHANGES_REQUESTED** [review]

File: index.js (hunk 3 above)
Comment (penx) [reply index.js]: Thanks, I'd meant to check on the best way to do this. I'm hoping this is the correct way:

https://github.com/fastify/fastify-reply-from/pull/141/commits/45bbd9da38389be8a84ff37f0c0df77902d08c6c

File: index.js [hunk 4]
```diff
@@ -29,6 +29,9 @@ module.exports = fp(function from (fastify, opts, next) {
     sessionTimeout: opts.sessionTimeout,
     undici: opts.undici
   })
+
+  fastify.setErrorHandler((error, request, reply) => errorHandler(error, request, reply, onErrorDefault))
```
Comment (mcollina) [inline index.js:33]: unfortunately this is not viable either. This will be implemented by the developers using the plugin.

Comment (penx) [reply index.js:33]: are you saying that `fastify.setErrorHandler` may be used higher up in the context and this would override it?

Perhaps there is a way we can intercept and change the error but still pass it up to the parent context error handler?

Comment (mcollina) [reply index.js:33]: I don't think we should intercept that error.

Comment (mcollina) [reply index.js:33]: adding an error handler here will override any other that were previously set.

Comment (penx) [reply index.js]: It would be nice for the user to get a 504 rather than 500, but I agree it may be overkill to handle this at the moment.

I've reverted the last few commits, removed the redundant check for UND_ERR_BODY_TIMEOUT and added a test.

**Review by mcollina — APPROVED** [review]
lgtm
